File: lib/rgfa/field/sequence_gfa2.rb
module RGFA::Field::SequenceGFA2

  def unsafe_decode(string)
    if string.placeholder?
      return RGFA::Placeholder.new
    else
      return string
    end
  end

  def decode(string)
    object = unsafe_decode(string)
    validate_decoded(object)
    return object
  end

  def validate_encoded(string)
    if string !~ /^[!-~]+$/
      raise RGFA::FormatError,
        "the string #{string.inspect} is not a valid GFA2 sequence\n"+
        "(it contains spaces and/or non-printable characters)"
    end
  end

  def validate_decoded(object)
    case object
    when RGFA::Placeholder
    when String
      validate_encoded(object)
    else
      raise RGFA::TypeError,
        "the class is incompatible with the datatype\n"+
        "(accepted classes: String, RGFA::Placeholder)"
    end
  end

  def unsafe_encode(object)
    object.to_s
  end

  def encode(object)
    case object
    when RGFA::Placeholder
      return object.to_s
    when String
      validate_encoded(object)
      return object
    else
      raise RGFA::TypeError,
        "the class is incompatible with the datatype\n"+
        "(accepted classes: String, RGFA::Placeholder)"
    end
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/sequence_gfa2.py
import gfapy
import re

def unsafe_decode(string):
  if string == "*":
    return gfapy.Placeholder()
  else:
    return string

def decode(string):
  obj = unsafe_decode(string)
  validate_decoded(obj)
  return obj

def validate_encoded(string):
  if not re.match(r"^[!-~]+$", string):
    raise gfapy.FormatError(
      "the string {} is not a valid GFA2 sequence\n".format(repr(string))+
      "(it contains spaces and/or non-printable characters)")

def validate_decoded(obj):
  if isinstance(obj, gfapy.Placeholder):
    pass
  elif isinstance(obj, str):
    validate_encoded(obj)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, gfapy.Placeholder)")

def unsafe_encode(obj):
  return str(obj)

def encode(obj):
  if isinstance(obj, gfapy.Placeholder):
    return str(obj)
  elif isinstance(obj, str):
    validate_encoded(obj)
    return obj
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, gfapy.Placeholder)")
